### DISCUSSION TITLE
Populate does not work on entity which has a default filter

### DISCUSSION THREAD

**Discussion by decline** [mikro-orm/mikro-orm#4840] [General]
**Describe the bug**

I have a Filter which is enabled by default (to query for a specific environment `dev` or `prod`). If the filter is not set in the query, it will filter by `env: null` by default:

```ts
export interface WithEnvArguments {
  env?: 'dev' | 'prod';
}

export const WithEnv = (): ClassDecorator =>
  Filter({
    name: 'env',
    cond: ({ env }: WithEnvArguments = {}) => {
      if (!env) {
        return {
          env: null
        };
      }
      return {
        env
      };
    },
    default: true
  });
```

Now I have an entity, that uses this filter:

```ts
@Entity()
@WithEnv()
export class SubscriptionPlanEntity {
  @PrimaryKey()
  name!: string;

  @Property()
  credits!: number;

  @Property()
  env!: string;

  [PrimaryKeyType]?: string;
}
```

This entity is part of a `@ManyToOne` relation in another entity, like here:

```ts
@Entity()
export class SubscriptionEntity {
  @PrimaryKey({ type: UuidType })
  id = v4();
  
  @ManyToOne(() => SubscriptionPlanEntity)
  subscriptionPlan!: SubscriptionPlanEntity;
}
```

**Issue:** It is not possible to populate `subscriptionPlan`, if you query for this entity with another value for `subscriptionPlan.env` (for example `'dev'`) then the default of the filter (which is `null`).

**To Reproduce**
1. Define entities as above
2. Query for one entity that has a different `env` as the default of the Filter and try to populate `subscriptionPlan`:

```ts
const entity = await this.subscriptionRepository.findOneOrFail(
  {
    subscriptionPlan: {
      env: 'dev'
    }
  },
  { populate: ['subscriptionPlan'] }
);
console.log(entity);
```

It will not populate the `subscriptionPlan` field:

```ts
SubscriptionEntity {
  id: '82d54ed3-d7d2-45d5-a4a8-a7cfe8fafdc5',
  subscriptionPlan: (SubscriptionPlanEntity) { name: 'basic' },
}
```

3. Now if you remove the default behaviour of the filter, by setting `default: false` there and run the query again, it will populate the entity:

```ts
SubscriptionEntity {
  id: '82d54ed3-d7d2-45d5-a4a8-a7cfe8fafdc5',
  subscriptionPlan: SubscriptionPlanEntity {
    name: 'basic',
    credits: 20,
    env: 'dev'
  },
}
```

**Expected behavior**
It seems that the DB query, which is done to populate the entity (also) uses the default filter and is therefore not able to find the entity. I would expect that the populate query would not use any filters, since it should just load the required entity by it's foreign key.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.15.0 |
| typescript | 5.1.6 |
| mikro-orm | 5.7.13 |
| your-driver | postgres |

**Comment by B4nan** [comment]
> mikro-orm | 5.7.13

Please use the latest version before you report anything, you are quite behind. What you are describing here sounds like a few other reports that were fixed with v5.8.9

If that won't help, provide a complete reproduction (but without nest - repros should be minimal, no webservers are needed if its an ORM problem). 

> It seems that the DB query, which is done to populate the entity (also) uses the default filter and is therefore not able to find the entity. I would expect that the populate query would not use any filters, since it should just load the required entity by it's foreign key.

Filters are just a way to alter the query, they are not special anyhow, you can consider them the same as if you would specify the conditions in the query directly. As such, they are not applied to the populate queries (there is `populateWhere` option to control this). 

My guess is you are using joined strategy (next time provide the ORM config right ahead! :]) and you are simply too behind the fixes, I remember fixing similar things in one of the v5.8 releases.

**Comment by decline** [comment]
Hi @B4nan, thank you for your quick reply!

I updated my `@mikro-orm/*` packages to this latest version, but still it did not solve the issue.

I created a minimal repository that re-produces this issue as you requested here: https://github.com/decline/mikro-orm-issue

Description in README should hopefully be enough to get it running, please let me know if you're facing any issues.

**Reply by B4nan** [reply]
I don't see anything wrong in there, you define a global filter and it is applied when loading the target entity. Populate queries are not special anyhow, they will respect the filters. The filter you set results in the child entity being not found, that is why it's not populated.

```
[query] select "s0".* from "subscription_entity" as "s0" where "s0"."id" = '1' limit 1 [took 3 ms, 1 result]
[query] select "s0".* from "subscription_plan_entity" as "s0" where "s0"."env" is null and "s0"."name" in ('for-dev-only') order by "s0"."name" asc [took 2 ms, 0 result]
This will NOT populate SubscriptionEntity {
  id: '1',
  subscriptionPlan: (SubscriptionPlanEntity) { name: 'for-dev-only' }
}
[query] select "s0".* from "subscription_entity" as "s0" where "s0"."id" = '2' limit 1 [took 1 ms, 1 result]
[query] select "s0".* from "subscription_plan_entity" as "s0" where "s0"."env" is null and "s0"."name" in ('for-all') order by "s0"."name" asc [took 1 ms, 1 result]
This WILL populate SubscriptionEntity {
  id: '2',
  subscriptionPlan: SubscriptionPlanEntity { name: 'for-all', credits: 20, env: null }
}
```

Here you can see the `"s0"."env" is null` part which is from your filter, applied the same way to both queries - but in your data the plan has the `env`. You can see the first populate query returning 0 results.

**Reply by B4nan** [reply]
You can add `filters: false` to the find options to disable the filters.

> It seems that the DB query, which is done to populate the entity (also) uses the default filter and is therefore not able to find the entity. I would expect that the populate query would not use any filters, since it should just load the required entity by it's foreign key.

This expectation is indeed wrong. One of the use cases for global filters is implementing soft deletes - imagine you would populate some soft deleted entity, and it would end up in the results just because you loaded it via `em.populate` instead of `em.find`... Those things are just different ways for doing the same.

**Reply by decline** [reply]
Ok, thank you so much! Your example with a soft-delete filter (which I also have in use, btw.) definitely makes sense. Now I can also see, that this is indeed the correct behaviour. Will keep it in mind when working with filters.

Also, kudos to your work on mikro-orm, definitely the best TypeScript ORM I've worked with so far!
